pkg/controller/statusmanager/pod_status.go

```diff
@@ -175,54 +177,67 @@ func (status *StatusManager) updateNodeStatus(nodeName string, ready bool, reaso
-						messages = messages + fmt.Sprintf("nsx-node-agent/%s not running: %s. ", containerStatus.Name, containerStatus.State.Terminated.Reason)
-					} else {
-						messages = messages + fmt.Sprintf("nsx-node-agent/%s not running: %s. ", containerStatus.Name, containerStatus.State.Terminated.Message)
-					}
+	containerStatus := corev1.ContainerStatus{}
+	ready := true
+	// messages is used to store the error message from conatiner status,
+	// tmpMsgs is for the status that pod just restarted in a short time
+	var messages, tmpMsgs string
+	nodeName := pod.Spec.NodeName
+	var startedAt *time.Time
+	var lastStartedAt *time.Time
+	// when pod is pending, its ContainerStatuses is empty
+	if len(pod.Status.ContainerStatuses) == 0 {
+		ready = false
+		messages = fmt.Sprintf("nsx-node-agent not running: %s.", pod.Status.Phase)
+	}
+	for _, containerStatus = range pod.Status.ContainerStatuses {
+		if containerStatus.State.Running == nil {
+			ready = false
+			if containerStatus.State.Waiting != nil {
+				if containerStatus.State.Waiting.Message == "" {
+					messages = messages + fmt.Sprintf("%s/%s not running: %s. ", pod.Name, containerStatus.Name, containerStatus.State.Waiting.Reason)
+				} else {
+					messages = messages + fmt.Sprintf("%s/%s not running: %s. ", pod.Name, containerStatus.Name, containerStatus.State.Waiting.Message)
 				}
-
+			} else if containerStatus.State.Terminated != nil {
+				if containerStatus.State.Terminated.Message == "" {
+					messages = messages + fmt.Sprintf("%s/%s not running: %s. ", pod.Name, containerStatus.Name, containerStatus.State.Terminated.Reason)
+				} else {
+					messages = messages + fmt.Sprintf("%s/%s not running: %s. ", pod.Name, containerStatus.Name, containerStatus.State.Terminated.Message)
+				}
+			}
+		} else {
+			// pod is running, but we should check if it just restarted
+			startedAt = &containerStatus.State.Running.StartedAt.Time
+			// there're 3 containers in nsx-node-agent pod, we should check the latest started time
+			if lastStartedAt == nil || (*startedAt).Sub(*lastStartedAt) > 0 {
+				lastStartedAt = startedAt
+			}
+			runningSeconds := int64(now.Sub(*startedAt) / 1000000000)
+			if runningSeconds < operatortypes.TimeBeforeRecoverNetwork {
+				log.Info(fmt.Sprintf("NSX node agent for node %s started before %d seconds", nodeName, runningSeconds))
+				tmpMsgs = tmpMsgs + fmt.Sprintf("%s/%s restarted before %d seconds. ", pod.Name, containerStatus.Name, runningSeconds)
 			}
 		}
-		status.setNodeNetworkUnavailable(nodeName, ready, "NSXNodeAgent", messages, sharedInfo)
 	}
+	if len(tmpMsgs) > 0 {
+		// if the pod just restarted, we still set the ready to false then try to set it to true,
+		// i.e. invoke the setNodeNetworkUnavailable 2 times because there's no sequent pod status change to
+		// trigger the pod controller to set ready to true.
+		status.setNodeNetworkUnavailable(nodeName, false, nil, "NSXNodeAgent", messages + tmpMsgs, sharedInfo)
```

Comment:
will this have the downside of requiring extra time for the node to become ready the first time the agent starts?

Reply:
Yes, when the user setup nsx-node-agent on a node at the first time, the node will be NetworkUnavailable=True then take 3 minutes to be observed, then become NetworkUnavailable=False. I don't think the first boot should have privilege to become healthy without waiting... Do you think we should have a special process for the first time case?

Reply:
how about we mark NetworkUnavailable as True if 2 or more successive failures are observed (so 1 is ok), or if the node doesn't come into Running status in under a minute?

Reply:
> how about we mark NetworkUnavailable as True if 2 or more successive failures are observed (so 1 is ok), or if the node doesn't come into Running status in under a minute?

Sorry, I didn't understand the goal of your suggestion. This patch is to solve the problem that the pod is still scheduled to the node which nsx-node-agent shows running but in fact not healthy, the nsx-node-agent will be running for about 3 minutes then restart, like this:
```
# kubectl get pod -nnsx-system -owide -w
NAME                      READY   STATUS    RESTARTS   AGE    IP           NODE                   NOMINATED NODE   READINESS GATES
nsx-ncp-7558558bb-6hz95   1/1     Running   0          127m   172.10.0.4   sc2-hs2b-dhcp-70-114   <none>           <none>
nsx-ncp-bootstrap-nlbzz   1/1     Running   0          127m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
nsx-ncp-bootstrap-rw4zg   1/1     Running   0          127m   172.10.0.2   sc2-hs2b-dhcp-70-197   <none>           <none>
nsx-ncp-bootstrap-zz9xf   1/1     Running   0          127m   172.10.0.4   sc2-hs2b-dhcp-70-114   <none>           <none>
nsx-node-agent-75g27      3/3     Running   0          127m   172.10.0.4   sc2-hs2b-dhcp-70-114   <none>           <none>
nsx-node-agent-7r672      3/3     Running   0          127m   172.10.0.2   sc2-hs2b-dhcp-70-197   <none>           <none>
nsx-node-agent-j5qsc      3/3     Running   26         127m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
nsx-node-agent-j5qsc      3/3     Running   27         128m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
nsx-node-agent-j5qsc      2/3     CrashLoopBackOff   27         131m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
nsx-node-agent-j5qsc      3/3     Running            28         136m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
nsx-node-agent-j5qsc      3/3     Running            29         139m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
nsx-node-agent-j5qsc      2/3     CrashLoopBackOff   29         142m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
nsx-node-agent-j5qsc      3/3     Running            30         147m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
nsx-node-agent-j5qsc      3/3     Running            31         149m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
nsx-node-agent-j5qsc      2/3     CrashLoopBackOff   31         152m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
nsx-node-agent-j5qsc      3/3     Running            32         157m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
nsx-node-agent-j5qsc      3/3     Running            33         160m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
nsx-node-agent-j5qsc      2/3     CrashLoopBackOff   33         163m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
nsx-node-agent-j5qsc      3/3     Running            34         168m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
nsx-node-agent-j5qsc      3/3     Running            35         171m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
nsx-node-agent-j5qsc      2/3     CrashLoopBackOff   35         174m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
nsx-node-agent-j5qsc      3/3     Running            36         179m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
nsx-node-agent-j5qsc      3/3     Running            37         3h2m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
nsx-node-agent-j5qsc      2/3     CrashLoopBackOff   37         3h5m   172.10.0.3   sc2-hs2b-dhcp-71-67    <none>           <none>
```
in the above case, e.g. during 136m ~ 141m, although the nsx-node-agent-j5qsc is in "Running", the node still should have NetworkUnavailable=True.

Reply:
Ok, went over the bug again. What you are doing seems right. But in this example, 180 sec wait time won’t help either, right?

Reply:
I think 180 sec is enough. From the following output, we can see the time duration between 2 restarts is 2m43s and 2m51s, which are less than 3min. In my test didn't see the time duration > 3min. I think the patch can handle the following case (keep the NetworkUnavailable=true)
```
# kubectl get pods -n nsx-system -o wide -w nsx-node-agent-flrd5
NAME                   READY   STATUS    RESTARTS   AGE   IP           NODE                          NOMINATED NODE   READINESS GATES
nsx-node-agent-flrd5   3/3     Running   0          20s   172.10.0.4   wdc-rdops-vm09-dhcp-224-189   <none>           <none>
nsx-node-agent-flrd5   3/3     Running   1          2m43s   172.10.0.4   wdc-rdops-vm09-dhcp-224-189   <none>           <none>
nsx-node-agent-flrd5   3/3     Running   2          5m34s   172.10.0.4   wdc-rdops-vm09-dhcp-224-189   <none>           <none>
```

Reply:
got it. thanks!